Add api to register setup checks

CarlSchwan opened the pull request:
Add an API for applications to register setup checks.
Rework setup checks UI.

CarlSchwan committed:
Add api to register setup checks

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

CarlSchwan committed:
Progress

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

kesselb commented:
I like the idea :+1: 

Hopefully, we manage to finish this time https://github.com/nextcloud/server/pull/22527 :see_no_evil:

come-nc committed:
Fix SetupChecks/LdapInvalidUuids.php

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix UI and add PhpOutdated check

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix docblock and types for new public API

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Small cleanups in SetupCheck classes

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix lint problems in vue files

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Move existing setup checks to new API

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
/compile /

come-nc committed:
Fix PHPDoc in OCP for new SetupResult class

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Migrate NeedsSystemAddressBookSync to new ISetupCheck API

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Migrate ReadOnlyConfig and DefaultPhoneRegionSet to new API

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

kesselb commented on apps/settings/lib/SetupChecks/NeedsSystemAddressBookSync.php:
```diff
@@ -28,19 +29,29 @@
-		return $this->config->getAppValue('dav', 'needs_system_address_book_sync', 'no') === 'no';
+	public function run(): SetupResult {
+		if ($this->config->getAppValue('dav', 'needs_system_address_book_sync', 'no') === 'no') {
+			return new SetupResult(SetupResult::SUCCESS, $this->l10n->t('The address book sync has already run'));
```
```suggestion
			return new SetupResult(SetupResult::SUCCESS, $this->l10n->t('No outstanding DAV system address book sync.'));
```

come-nc committed:
Fix description for SetupChecks/NeedsSystemAddressBookSync

Co-authored-by: Daniel <mail@danielkesselberg.de>
Signed-off-by: Côme Chilliet <91878298+come-nc@users.noreply.github.com>

come-nc committed:
Merge setupchecks from new API into old UI

This will allow migrating checks to the new API without touching the UI
 for now.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Add missing licences and copyright in SetupCheck

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Migrate InternetConnectivity check to new API

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Add OCP interface for SetupCheckManager

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix tests for setup checks

Had to remove tests for migrated checks, we should add tests for each
 SetupCheck class later on.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Remove UI draft

For now we will keep the existing UI and only change the API behind it

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

kesselb commented on apps/settings/lib/SetupChecks/NeedsSystemAddressBookSync.php:
```diff
@@ -28,19 +29,29 @@
+use OCP\SetupCheck\ISetupCheck;
+use OCP\SetupCheck\SetupResult;
+
+class NeedsSystemAddressBookSync implements ISetupCheck {
```
The check is registered in the dav app, and therefore it would be nice to move it over there if that's already possible.

kesselb reviewed (approved):
Very nice, thanks for taking care :+1:

come-nc committed:
Add debug logging when running setup checks to help with debugging slow results

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Move NeedsSystemAddressBookSync to dav application where it’s registered

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

Altahrim commented on lib/public/SetupCheck/SetupResult.php:
```diff
@@ -0,0 +1,94 @@
+	 * @param self::SUCCESS|self::INFO|self::WARNING|self::ERROR $severity
+	 * @since 28.0.0
+	 */
+	public function __construct(
```
I think we are not used to it at Nextcloud, but what about named constructors?

```php
public static function success(?string $description = null, ?string $linkToDoc = null): self {
	return new self(self::SUCCESS, $description, $linkToDoc);
}

public static function info(?string $description = null, ?string $linkToDoc = null): self {
	return new self(self::INFO, $description, $linkToDoc);
}

…

private function __construct(
		private string $severity,
		private ?string $description = null,
		private ?string $linkToDoc = null,
) {
}
```

It's a bit more verbose here, but usage is simpler:
```php
return SetupResult::success('We made it!')
```

come-nc replied:
I’m not sure I want this. Can be added later, no?
Do you mean `__construct` would be `private` in this case? (which means if we do it we must do it before 28 is out)

ChristophWurst replied:
It's used for `\OC\Authentication\Login\LoginResult::success`, `\OC\Talk\ConversationOptions::default` and `\OCP\Search\SearchResult::complete`. I think the pattern is nice for enum-like objects where you want to control the type of instances that can be created.

ChristophWurst commented on lib/public/SetupCheck/ISetupCheckManager.php:
```diff
@@ -0,0 +1,37 @@
+	 * @since 28.0.0
+	 * @return array<string,array<string,SetupResult>> Result of each check, first level key is category, second level key is title
+	 */
+	public function runAll(): array;
```
This doesn't need to be exposed, does it? Apps won't run the checks. The server runs them.

come-nc replied:
https://github.com/nextcloud/admin_overview will run the checks.
And `settings` runs the checks, which is technically an app.
And I do hope at some point to add an `occ` command to run the checks from cli, and maybe one day include the check results into the system report, which means `support` app may run the checks.

come-nc replied:
(Some of those apps will run the checks through the rest API and not `OCP` though, but still)

come-nc commented on apps/settings/lib/SetupChecks/InternetConnectivity.php:
```diff
@@ -0,0 +1,100 @@
+			return new SetupResult(SetupResult::SUCCESS, $this->l10n->t('Internet connectivity is disabled in configuration file.'));
+		}
+
+		$siteArray = $this->config->getSystemValue('connectivity_check_domains', [
```
Yeah we’d need a getSystemValueArray like what we have for bool/int/string but it was not added yet.
For the current PR this code is only moved around it was already like this.

come-nc committed:
Remove useless var in apps/settings/lib/SetupChecks/CheckUserCertificates.php

Co-authored-by: Benjamin Gaussorgues <benjamin.gaussorgues@nextcloud.com>
Signed-off-by: Côme Chilliet <91878298+come-nc@users.noreply.github.com>

come-nc committed:
Change SetupResult API to named constructors

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

Altahrim commented on apps/settings/lib/SetupChecks/DefaultPhoneRegionSet.php:
```diff
@@ -0,0 +1,59 @@
+		if ($this->config->getSystemValueString('default_phone_region', '') !== '') {
+			return SetupResult::success($this->config->getSystemValueString('default_phone_region', ''));
+		} else {
+			return new SetupResult(
+				SetupResult::INFO,
+				$this->l10n->t('Your installation has no default phone region set. This is required to validate phone numbers in the profile settings without a country code. To allow numbers without a country code, please add "default_phone_region" with the respective ISO 3166-1 code of the region to your config file.'),
+				'https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2#Officially_assigned_code_elements'
+			);
```
```suggestion
			return SetupResult::info(
				$this->l10n->t('Your installation has no default phone region set. This is required to validate phone numbers in the profile settings without a country code. To allow numbers without a country code, please add "default_phone_region" with the respective ISO 3166-1 code of the region to your config file.'),
				'https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2#Officially_assigned_code_elements'
			);
```

Altahrim commented on lib/public/SetupCheck/SetupResult.php:
```diff
@@ -0,0 +1,127 @@
+		private ?string $description = null,
+		private ?string $linkToDoc = null,
+	) {
+		$this->severity = $severity;
+		$this->description = $description;
+		$this->linkToDoc = $linkToDoc;
```
```suggestion
```

Altahrim commented on lib/private/SetupCheck/SetupCheckManager.php:
```diff
@@ -0,0 +1,58 @@
+			$this->logger->debug('Running check '.get_class($setupCheckObject));
+			$setupResult = $setupCheckObject->run();
+			$category = $setupCheckObject->getCategory();
+			if (!isset($results[$category])) {
+				$results[$category] = [];
+			}
```
Nit: 
I think it can be:
```suggestion
			$results[$category] ??= [];
```

come-nc committed:
Small code style fix in SetupCheckManager

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix jsunit tests for SetupChecks

Tests could be simplified to test generic setup check directly with fake
 data instead of trying to mimic real steps results but for now this
 will do.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

nickvergessen commented:
Docs would be cool on this one 😎
